File: src/main/java/com/uwetrottmann/trakt/v2/services/Recommendations.java
package com.uwetrottmann.trakt.v2.services;

import com.uwetrottmann.trakt.v2.exceptions.OAuthUnauthorizedException;
import com.uwetrottmann.trakt.v2.entities.Movie;
import com.uwetrottmann.trakt.v2.entities.Show;
import com.uwetrottmann.trakt.v2.enums.Extended;
import retrofit.client.Response;
import retrofit.http.DELETE;
import retrofit.http.EncodedQuery;
import retrofit.http.GET;
import retrofit.http.Path;

import java.util.List;

public interface Recommendations {

    /**
     * <b>OAuth Required</b>
     *
     * <p> Personalized movie recommendations for a user. Results returned with the top recommendation first.
     */
    @GET("/recommendations/movies")
    List<Movie> movies(
            @EncodedQuery("extended") Extended extended
    ) throws OAuthUnauthorizedException;

    /**
     * <b>OAuth Required</b>
     *
     * <p> Dismiss a movie from getting recommended anymore.
     *
     * @param movieId trakt ID, trakt slug, or IMDB ID. Example: "tron-legacy-2010".
     */
    @DELETE("recommendations/movies/{id}")
    Response dismissMovie(
            @Path("id") String movieId
    ) throws OAuthUnauthorizedException;

    /**
     * <b>OAuth Required</b>
     *
     * <p> Personalized show recommendations for a user. Results returned with the top recommendation first.
     */
    @GET("/recommendations/shows")
    List<Show> shows(
            @EncodedQuery("extended") Extended extended
    ) throws OAuthUnauthorizedException;

    /**
     * <b>OAuth Required</b>
     *
     * <p> Dismiss a show from getting recommended anymore.
     *
     * @param showId trakt ID, trakt slug, or IMDB ID. Example: 922.
     */
    @DELETE("recommendations/shows/{id}")
    Response dismissShow(
            @Path("id") String showId
    ) throws OAuthUnauthorizedException;

}

File: src/main/kotlin/com/redissi/trakt/services/Recommendations.kt
package com.redissi.trakt.services

import com.redissi.trakt.entities.Movie
import com.redissi.trakt.entities.Show
import com.redissi.trakt.enums.Extended
import retrofit2.Response
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query

interface Recommendations {
    /**
     * **OAuth Required**
     *
     *
     *  Personalized movie recommendations for a user. Results returned with the top recommendation first.
     *
     * @param page Number of page of results to be returned. If `null` defaults to 1.
     * @param limit Number of results to return per page. If `null` defaults to 10.
     */
    @GET("recommendations/movies")
    suspend fun movies(
        @Query("page") page: Int?,
        @Query("limit") limit: Int?,
        @Query(value = "extended", encoded = true) extended: Extended?
    ): List<Movie>?

    /**
     * **OAuth Required**
     *
     *
     *  Dismiss a movie from getting recommended anymore.
     *
     * @param movieId trakt ID, trakt slug, or IMDB ID. Example: "tron-legacy-2010".
     */
    @DELETE("recommendations/movies/{id}")
    suspend fun dismissMovie(
        @Path("id") movieId: String?
    ): Response<Unit>

    /**
     * **OAuth Required**
     *
     *
     *  Personalized show recommendations for a user. Results returned with the top recommendation first.
     *
     * @param page Number of page of results to be returned. If `null` defaults to 1.
     * @param limit Number of results to return per page. If `null` defaults to 10.
     */
    @GET("recommendations/shows")
    suspend fun shows(
        @Query("page") page: Int?,
        @Query("limit") limit: Int?,
        @Query(value = "extended", encoded = true) extended: Extended?
    ): List<Show>?

    /**
     * **OAuth Required**
     *
     *
     *  Dismiss a show from getting recommended anymore.
     *
     * @param showId trakt ID, trakt slug, or IMDB ID. Example: 922.
     */
    @DELETE("recommendations/shows/{id}")
    suspend fun dismissShow(
        @Path("id") showId: String?
    ): Response<Unit>
}